File: sol/10.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        ns = len(s)
        np = len(p)
        memo = [[False] * (np + 1) for _ in range(ns + 1)]
        memo[-1][-1] = True
        for i in range(ns, -1, -1):
            for j in range(np - 1, -1, -1):
                # this is usually a good optimization,
                # but some test cases has patten with consecutive '*'
                # if p[j] == '*':
                #     continue
                m = i != ns and (p[j] == '.' or p[j] == s[i])
                if j < np - 1 and p[j + 1] == '*':
                    memo[i][j] = memo[i][j + 2] or (m and memo[i + 1][j])
                else:
                    memo[i][j] = m and memo[i + 1][j + 1]
        return memo[0][0]

File: sol/10/10.cpp
struct Solution {
    bool isMatch(string &s, string &p) {
        int ns = s.size(), np = p.size();
        vector<vector<bool>> memo(ns + 1, vector<bool>(np + 1, false));
        memo[ns][np] = true;
        for (int i = ns; i >= 0; --i){
            for (int j = np - 1; j >= 0; --j){
                bool m = i != ns && (p[j] == '.' || p[j] == s[i]);
                if (j < np - 1 && p[j + 1] == '*'){
                    memo[i][j] = memo[i][j + 2] || (m && memo[i + 1][j]);
                } else {
                    memo[i][j] = m && memo[i + 1][j + 1];
                }
            }
        }
        return memo[0][0];
    }
};
